### PR TITLE
Kernel info

### PR DESCRIPTION
**Pull request by rgbkrk**
In an effort to surface the right information in the status bar in the kernel, I went after cleaning up our kernel info and making the notebook app take a `contentRef`.

## `setKernelInfo` action and `KernelInfo` record

We had an action called `setKernelInfo` that was only used on Desktop (as far as I can tell) that I've renamed to `setKernelspecInfo`. I've done this so we can have a `setKernelInfo` that actually matches the `kernel_info_request` and `kernel_info_reply` pairs.

There are now proper types for `KernelInfo`, and I've created a flattened version of the `kernel_info_reply` content for us to use.

## `<NotebookApp />`

In the process here, I've made the `<NotebookApp />` component require a `contentRef`. On Desktop it uses the `contentRef` set at the start (since we never swap it out on the same window).

## `<StatusBar />`

`<StatusBar />` is now a connected component that accepts props for `contentRef` and `kernelRef` so that it can look up exactly the information it needs.

## Generic import cleanups

In the process of doing this PR I ended up doing some cleanup of imports from core to simplify how I was working on this.

### CODE REVIEW

File: packages/core/src/state/entities/kernel-info.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  // (e.g. IPython.__version__ for the IPython kernel)
+  implementationVersion: string,
+
+  // A banner of information about the kernel,
+  // which may be displayed in console environments.
+  banner: string,
+
+  // Optional: A list of dictionaries, each with keys 'text' and 'url'.
+  // These can be displayed in the help menu for the UI.
+  help_links?: Array<{ text: string, url: string }>,
```
Comment (theengineear) [inline packages/core/src/state/entities/kernel-info.js]: Should this be Immutable here?

File: packages/core/src/state/entities/kernel-info.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  // Nbconvert exporter, if notebooks written with this kernel should
+  // be exported with something other than the general 'script'
+  // exporter.
+  nbconvert_exporter: string
+};
+
+export type KernelInfoRecord = Immutable.RecordOf<KernelInfo>;
+
+export function makeKernelInfoRecord(kernelInfo: KernelInfo): KernelInfoRecord {
```
Comment (theengineear) [inline packages/core/src/state/entities/kernel-info.js]: Nice, trying out the wrapper version of this?

Comment (rgbkrk) [reply packages/core/src/state/entities/kernel-info.js]: Yeah, works pretty well and forces these types to all be set and not have to do a big `null` hunt. The only strange thing is our creation of records. I wonder if I should make sure that two equivalent set of props passed into would evalute `Immutable.is(a,b)` to `true`.

File: applications/notebook-on-next/pages/edit.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
-// @flow
+// FIXME: To make it easier to iterate on supported apps, flow is disabled for
```
Comment (rgbkrk) [inline applications/notebook-on-next/pages/edit.js:1]: I've fully disabled notebook-on-next until we have time/resources to work on it again.

File: packages/core/__tests__/epics/kernel-lifecycle-spec.js [hunk 4]
```diff
@@ -26,21 +26,123 @@ describe("acquireKernelInfo", () => {
...
+            fileExtension: ".py",
+            pygmentsLexer: "ipython3",
+            codemirrorMode: { name: "ipython", version: 3 },
+            nbconvertExporter: "python"
+          }
+        }
+      }
+    ]);
+
+    done();
```
Comment (rgbkrk) [inline packages/core/__tests__/epics/kernel-lifecycle-spec.js:145]: Now with real tests (compared to what this did before)!

File: packages/core/src/components/notebook-app.js [hunk 5]
```diff
@@ -354,33 +339,32 @@ const mapStateToProps = (
     );
   }
 
-  let kernel = {
-    kernelSpecName: null,
-    status: null
-  };
+  // TODO: Determine and fix things so we have one reliable place for the kernelRef
+  const kernelRef =
+    selectors.currentKernelRef(state) || ownProps.kernelRef || model.kernelRef;
```
Comment (rgbkrk) [inline packages/core/src/components/notebook-app.js:344]: In particular, I noticed I couldn't reliably rely on `model.kernelRef` being set and didn't want to add on the reducers refactor here (yet).

File: packages/core/src/components/notebook-app.js [hunk 6]
```diff
@@ -332,18 +321,14 @@ const mapStateToProps = (
...
       theme: selectors.userPreferences(state).theme,
-      lastSaved: content.lastSaved,
       cellOrder: Immutable.List(),
-      // TODO: deal with current kernel ref
-      kernelStatus: "unknown",
-      languageDisplayName: "unknown",
       transforms: ownProps.transforms || transforms,
       displayOrder: ownProps.displayOrder || displayOrder,
       codeMirrorMode: Immutable.Map({ name: "text/plain" }),
+      kernelRef: null,
```
Comment (rgbkrk) [inline packages/core/src/components/notebook-app.js:331]: Is there a pattern for returning an alternate component if we can't create the suite of props that we need to in `mapStateToProps`?

File: packages/core/src/state/entities/kernel-info.js (hunk 2 above)
Comment (rgbkrk) [reply packages/core/src/state/entities/kernel-info.js]: 😭 

```
> Immutable.is(Immutable.Record({a: 2})(), Immutable.Record({a: 2})())
false
```

File: packages/core/src/components/notebook-app.js (hunk 6 above)
Comment (theengineear) [reply packages/core/src/components/notebook-app.js:331]: Other than checking in the render method of _this_ component of the _parent_ component, I don't think there's a better pattern (if I'm understanding your question correctly).

File: packages/core/src/reducers/core/entities/contents.js [hunk 7]
```diff
@@ -878,8 +878,8 @@ function document(
       return updateCellStatus(state, action);
     case actionTypes.SET_LANGUAGE_INFO:
       return setLanguageInfo(state, action);
-    case actionTypes.SET_KERNEL_INFO:
-      return setKernelInfo(state, action);
+    case actionTypes.SET_KERNELSPEC_INFO:
```
Comment (theengineear) [inline packages/core/src/reducers/core/entities/contents.js:881]: Nice, thanks for cleaning these all up as well.

File: packages/core/src/epics/kernel-lifecycle.js [hunk 8]
```diff
@@ -86,7 +86,14 @@ export function acquireKernelInfo(
           langInfo: msg.content.language_info,
           kernelRef,
           contentRef
-        })
+        }),
+        {
+          type: "ACTUAL_KERNEL_INFO",
```
Comment (theengineear) [inline packages/core/src/epics/kernel-lifecycle.js]: 😁

File: applications/desktop/src/notebook/epics/github-publish.js [hunk 9]
```diff
@@ -1,8 +1,12 @@
...
 import { shell } from "electron";
 
-import type { ContentRef } from "@nteract/core/src/state";
-import { selectors, actions, actionTypes } from "@nteract/core";
+import {
+  selectors,
+  actions,
+  actionTypes,
+  state as stateModule
+} from "@nteract/core";
```
Comment (theengineear) [inline applications/desktop/src/notebook/epics/github-publish.js:9]: 💯 for standardizing these!

File: packages/core/src/components/notebook-app.js (hunk 6 above)
Comment (rgbkrk) [reply packages/core/src/components/notebook-app.js:331]: Seems like we'd have to have some wrapper component that delegates to a component based on whether we have "real" `content` and `model`. Something likely called `<NotebookApp />` as the export still, yet has its own `mapStateToProps` to basically dictate if we're showing a placeholder or the real notebook app. 🤷‍♂️ I'll get something better written up for this.

File: packages/core/src/epics/kernel-lifecycle.js (hunk 8 above)
Comment (rgbkrk) [reply packages/core/src/epics/kernel-lifecycle.js]: I should really learn better that I shouldn't have stray gunk in my commits so you don't see me fooling with data. 😄

File: packages/core/src/state/entities/kernel-info.js (hunk 1 above)
Comment (rgbkrk) [reply packages/core/src/state/entities/kernel-info.js]: Tackled this one by the way.

File: packages/core/src/state/entities/kernel-info.js (hunk 2 above)
Comment (rgbkrk) [reply packages/core/src/state/entities/kernel-info.js]: I did switch this to doing the proper thing with records by the way.

Comment (theengineear) [reply packages/core/src/state/entities/kernel-info.js]: Possibly fixed? https://github.com/facebook/immutable-js/commit/0c5ca8f344908e3c5b106bae7b008948e6a9f859

File: packages/core/__tests__/epics/websocket-kernel-spec.js [hunk 10]
```diff
@@ -7,7 +7,7 @@ import { actions, state as stateModule, epics as coreEpics } from "../../src";
 import { toArray } from "rxjs/operators";
 
 describe("launchWebSocketKernelEpic", () => {
-  test("", async function() {
+  test("launches remote kernels", async function() {
```
Comment (theengineear) [inline packages/core/__tests__/epics/websocket-kernel-spec.js:10]: 🥇

File: packages/core/src/state/entities/kernel-info.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  implementationVersion: string,
+
+  // A banner of information about the kernel,
+  // which may be displayed in console environments.
+  banner: string,
+
+  // Optional: A list of dictionaries, each with keys 'text' and 'url'.
+  // These can be displayed in the help menu for the UI.
+  // SPEC: help_links
+  helpLinks?: Array<{ text: string, url: string }>,
```
Comment (theengineear) [inline packages/core/src/state/entities/kernel-info.js]: I think I left a comment on this before. But I didn't see the resolution. Should we have this as an Immutable `List`?

File: packages/core/src/epics/kernel-lifecycle.js [hunk 12]
```diff
@@ -80,23 +81,35 @@ export function acquireKernelInfo(
     ofMessageType("kernel_info_reply"),
     first(),
     mergeMap(msg => {
+      const c = msg.content;
+      const l = c.language_info;
```
Comment (theengineear) [inline packages/core/src/epics/kernel-lifecycle.js]: Better variable names?

File: packages/core/src/state/entities/kernel-info.js [hunk 13]
```diff
@@ -35,8 +47,11 @@ export type KernelInfo = {
 
   // Optional: A list of dictionaries, each with keys 'text' and 'url'.
   // These can be displayed in the help menu for the UI.
+  // NOTE: To make it easy to use the KernelInfo type as a plain object in
+  //       actions, we allow this type to be an array of plain objects or the
+  //       fully immutable variety
   // SPEC: help_links
-  helpLinks?: Array<{ text: string, url: string }>,
+  helpLinks?: Array<HelpLink> | Immutable.List<HelpLinkRecord>,
```
Comment (theengineear) [inline packages/core/src/state/entities/kernel-info.js]: Ah. OK, i see you handle it here. 🤔 Still trying to understand why we need the Array? If anything, maybe the correct thing to do is type the API response and then type the resulting state separately?

File: packages/core/src/reducers/core/entities/kernels.js [hunk 14]
```diff
@@ -50,10 +50,22 @@ const byRef = (
...
+      }
+      switch (typeof codemirrorMode) {
+        case "string":
+          // already set as we want it
+          break;
+        case "object":
+          codemirrorMode = Immutable.Map(codemirrorMode);
+          break;
+        default:
+          // any other case results in falling back to language name
```
Comment (theengineear) [inline packages/core/src/reducers/core/entities/kernels.js:66]: 👌

File: applications/desktop/src/notebook/index.js [hunk 15]
```diff
@@ -77,6 +77,8 @@ export default class App extends React.PureComponent<Object, Object> {
           <Styles>
             <MathJax.Context input="tex">
               <NotebookApp
+                // The desktop app always keeps the same contentRef in a browser window
```
Comment (theengineear) [inline applications/desktop/src/notebook/index.js:80]: 👍

File: packages/core/src/components/notebook-app.js [hunk 16]
```diff
@@ -285,10 +285,7 @@ type PureNotebookProps = {
   languageDisplayName?: string,
   kernelStatus?: string,
   codeMirrorMode?: string | Immutable.Map<string, *>,
-  // TODO: Once we're willing to do multi-contents views, we should require this
-  //       to be passed in
-  // TODO: Fill in more from the convo with Andrew here
-  contentRef?: ContentRef,
```
Comment (theengineear) [inline packages/core/src/components/notebook-app.js:291]: 🎉

File: packages/core/src/components/notebook-app.js (hunk 5 above)
Comment (theengineear) [reply packages/core/src/components/notebook-app.js:344]: Yah. I think we *want* it to be `model.kernelRef`, but I'm guessing there's a race condition for getting that to be set? I.e., we might need to get here before the content gets returned or something?

File: packages/core/src/components/status-bar.js [hunk 17]
```diff
@@ -1,14 +1,21 @@
 /* @flow */
 import React from "react";
+
+import { connect } from "react-redux";
+
 import distanceInWordsToNow from "date-fns/distance_in_words_to_now";
 
+import * as selectors from "../selectors";
+import type { AppState } from "../state";
+import type { ContentRef, KernelRef } from "../state/refs";
```
Comment (theengineear) [inline packages/core/src/components/status-bar.js]: You should be able to import all this at once from the top level state.

File: packages/core/src/components/status-bar.js [hunk 18]
```diff
@@ -62,3 +69,27 @@ export default class StatusBar extends React.Component<Props> {
...
+const mapStateToProps = (
+  state: AppState,
+  ownProps: { contentRef: ContentRef, kernelRef: ?KernelRef }
+): Props => {
+  const { contentRef, kernelRef } = ownProps;
+  const content = selectors.content(state, { contentRef });
+  const kernel = kernelRef ? selectors.kernel(state, { kernelRef }) : null;
+
+  const lastSaved =
+    content && content.lastSaved ? content.lastSaved : undefined;
```
Comment (theengineear) [inline packages/core/src/components/status-bar.js]: Minor, but it feels weird to set it to undefined here. Can we use `null`?

File: packages/core/src/components/status-bar.js [hunk 19]
```diff
@@ -62,3 +69,27 @@ export default class StatusBar extends React.Component<Props> {
...
+  const kernelStatus =
+    kernel && kernel.status ? kernel.status : "not connected";
+  const kernelSpecDisplayName =
+    kernel && kernel.kernelSpecName ? kernel.kernelSpecName : "kernel";
```
Comment (theengineear) [inline packages/core/src/components/status-bar.js]: 😁 i like the generic default 👍 .

File: packages/core/src/components/status-bar.js [hunk 20]
```diff
@@ -80,10 +82,18 @@ const mapStateToProps = (
...
-  const kernelStatus =
-    kernel && kernel.status ? kernel.status : "not connected";
-  const kernelSpecDisplayName =
-    kernel && kernel.kernelSpecName ? kernel.kernelSpecName : "kernel";
+
+  const kernelStatus = kernel && kernel.status ? kernel.status : NOT_CONNECTED;
+
+  let kernelSpecDisplayName = " ";
+  if (kernelStatus === NOT_CONNECTED) {
+    kernelSpecDisplayName = "no kernel";
```
Comment (theengineear) [inline packages/core/src/components/status-bar.js]: even better!

File: packages/core/src/state/entities/kernel-info.js [hunk 21]
```diff
@@ -94,9 +94,27 @@ export type KernelInfo = {
 
 export type KernelInfoRecord = Immutable.RecordOf<KernelInfo>;
 
+// NOTE: This is not exported as the "makeXRecord" for KernelInfo so that we
+//       don't have nullables and can rely on the proper types. The developer is
+//       required to pass a fully instantiated KernelInfo into
+//       makeKernelInfoRecord below.
```
Comment (theengineear) [inline packages/core/src/state/entities/kernel-info.js]: 👍 cool. i like it.

File: packages/core/src/state/entities/kernel-info.js (hunk 13 above)
Comment (rgbkrk) [reply packages/core/src/state/entities/kernel-info.js]: This is at the level of needing payload vs. record typed appropriately. I suppose if I made these generics instead that would help.

File: packages/core/src/components/notebook-app.js (hunk 16 above)
Comment (rgbkrk) [reply packages/core/src/components/notebook-app.js:291]: 💥

File: packages/core/src/components/status-bar.js (hunk 17 above)
Comment (rgbkrk) [reply packages/core/src/components/status-bar.js]: so true

**Pull request merged**